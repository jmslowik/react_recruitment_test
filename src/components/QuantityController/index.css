.quantity-controller {
  display: flex;
}
.quantity-text {
  padding-right: 5px;
}

.quantity {
  font-weight: bold;
}

.spinner {
  position: absolute;
  right: 30px;
}

.spinner span {
  font-size: 19px !important;
}
