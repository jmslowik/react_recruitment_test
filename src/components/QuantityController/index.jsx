import React from 'react';
import './index.css';

const QuantityText = ({ value, className }) => (
  <div className={className}>
    Obecnie masz
    {' '}
    <span className="quantity">{value}</span>
    {' '}
    sztuk produktu
  </div>
);

const QuantityController = ({
  quantity, min = NaN, max = NaN, isBlocked = false, onChange,
}) => {
  const isQuantityMin = quantity === 0 || (min >= 0 && quantity <= min);
  const isQuantityMax = max >= 0 && quantity >= max;

  const onAdd = () => {
    if (!isQuantityMax && !isBlocked) {
      onChange(quantity + 1);
    }
  };

  const onSubtract = () => {
    if (!isQuantityMin && !isBlocked) {
      onChange(quantity - 1);
    }
  };

  return (
    <div className="quantity-controller">
      <QuantityText className="quantity-text" value={quantity} />
      <button
        type="button"
        onClick={onSubtract}
        disabled={isBlocked || isQuantityMin}
      >
        -
      </button>
      <button
        type="button"
        onClick={onAdd}
        disabled={isBlocked || isQuantityMax}
      >
        +
      </button>
    </div>
  );
};

export default QuantityController;
