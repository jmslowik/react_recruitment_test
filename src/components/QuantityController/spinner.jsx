import React from 'react';
import { Spin } from 'antd';

const Spinner = ({ className }) => (
  <div className={className}>
    <Spin
      size="small"
    />
  </div>
);

export default Spinner;
