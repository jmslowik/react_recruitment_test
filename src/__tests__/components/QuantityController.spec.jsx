/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { findButton } from '@/__tests__/__utils__';

import QuantityController from '@/components/QuantityController';

describe('QuantityController', () => {
  const quantity = 3;
  const min = 1;
  const max = 5;
  const onChange = jest.fn();

  const props = {
    quantity,
    min,
    max,
    onChange,
  };

  beforeEach(() => {
    onChange.mockClear();
  });

  it('should render without throwing an error', () => {
    const wrapper = shallow(<QuantityController />);
    expect(wrapper.is('.quantity-controller')).toBeTruthy();
  });

  it('should contain \'-\' button', () => {
    const wrapper = shallow(<QuantityController />);
    expect(wrapper.find('button').contains('-')).toBeTruthy();
  });

  it('should contain \'+\' button', () => {
    const wrapper = shallow(<QuantityController />);
    expect(wrapper.find('button').contains('+')).toBeTruthy();
  });

  it('should contain text: \'Obecnie masz x sztuk produktu\'', () => {
    const wrapper = mount(<QuantityController quantity={quantity} />);
    expect(wrapper.find('.quantity-text').at(1).text()).toEqual(`Obecnie masz ${quantity} sztuk produktu`);
  });

  it('should emit onChange event with quantity-1 when \'-\' button clicked', () => {
    const wrapper = shallow(<QuantityController {...props} />);
    const button = findButton(wrapper, '-');

    button.simulate('click');

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(quantity - 1);
  });

  it('should emit onChange event with quantity+1 when \'+\' button clicked', () => {
    const wrapper = shallow(<QuantityController {...props} />);
    const button = findButton(wrapper, '+');

    button.simulate('click');

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(quantity + 1);
  });

  it('should not allow to subtract quantity below min value', () => {
    const wrapper = shallow(<QuantityController {...props} quantity={min} />);
    const button = findButton(wrapper, '-');

    button.simulate('click');

    expect(onChange).not.toHaveBeenCalled();
  });

  it('should not allow to subtract quantity below 0 value', () => {
    const wrapper = shallow(<QuantityController max={max} quantity={0} onChange={onChange} />);
    const button = findButton(wrapper, '-');

    button.simulate('click');

    expect(onChange).not.toHaveBeenCalled();
  });

  it('should not allow to add quantity above max value', () => {
    const wrapper = shallow(<QuantityController {...props} quantity={max} />);
    const button = findButton(wrapper, '+');

    button.simulate('click');

    expect(onChange).not.toHaveBeenCalled();
  });

  it('should allow add quantity when max is not defied', () => {
    const wrapper = shallow(
      <QuantityController min={min} quantity={quantity} onChange={onChange} />,
    );
    const button = findButton(wrapper, '+');

    button.simulate('click');

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(quantity + 1);
  });

  it('should allow subtract quantity when min is not defied', () => {
    const wrapper = shallow(
      <QuantityController max={max} quantity={quantity} onChange={onChange} />,
    );
    const button = findButton(wrapper, '-');

    button.simulate('click');

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(quantity - 1);
  });

  it('should disable buttons when \'isBlocked\'', () => {
    const wrapper = shallow(
      <QuantityController {...props} isBlocked />,
    );

    const buttonAdd = findButton(wrapper, '-');
    const buttonSubtract = findButton(wrapper, '+');

    expect(buttonAdd.prop('disabled')).toEqual(true);
    expect(buttonSubtract.prop('disabled')).toEqual(true);
  });

  it('should be false by default \'isBlocked\'', () => {
    const wrapper = shallow(<QuantityController />);

    const buttonAdd = findButton(wrapper, '-');
    const buttonSubtract = findButton(wrapper, '+');

    expect(buttonAdd.prop('disabled')).toEqual(false);
    expect(buttonSubtract.prop('disabled')).toEqual(false);
  });

  it('should render correctly', () => {
    const tree = renderer.create(<QuantityController {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
