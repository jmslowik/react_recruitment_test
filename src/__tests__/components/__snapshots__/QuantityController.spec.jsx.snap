// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityController should render correctly 1`] = `
<div
  className="quantity-controller"
>
  <div
    className="quantity-text"
  >
    Obecnie masz
     
    <span
      className="quantity"
    >
      3
    </span>
     
    sztuk produktu
  </div>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    -
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    +
  </button>
</div>
`;
